src/main/java/com/tenx/ms/retail/order/service/OrderService.java

```diff
@@ -0,0 +1,137 @@
+import com.tenx.ms.retail.order.rest.dto.Order;
+import com.tenx.ms.retail.order.rest.dto.OrderDetail;
+import com.tenx.ms.retail.order.util.OrderConverter;
+import com.tenx.ms.retail.order.util.OrderDetailStatusEnum;
+import com.tenx.ms.retail.product.domain.ProductEntity;
+import com.tenx.ms.retail.product.repository.ProductRepository;
+import com.tenx.ms.retail.stock.domain.StockEntity;
+import com.tenx.ms.retail.stock.repository.StockRepository;
+import com.tenx.ms.retail.store.domain.StoreEntity;
+import com.tenx.ms.retail.store.repository.StoreRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.stereotype.Service;
+
+import javax.validation.ConstraintViolationException;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * Created by goropeza on 28/08/16.
+ */
+@Service
+public class OrderService {
+    @Autowired
+    private OrderConverter orderConverter;
+    @Autowired
+    private OrderRepository orderRepository;
+    @Autowired
+    private OrderDetailRepository orderDetailRepository;
+    @Autowired
+    private ClientRepository clientRepository;
+    @Autowired
+    private StoreRepository storeRepository;
+    @Autowired
+    private StockRepository stockRepository;
+    @Autowired
+    private ProductRepository productRepository;
+
+    @SuppressWarnings("PMD")
+    public Order createOrder(Long storeId, Order order) {
+        ClientEntity client;
+        StoreEntity store;
+        ProductEntity product;
+
+        Optional<StoreEntity> resultStore = storeRepository.findById(storeId);
+        if (!resultStore.isPresent()) {
+            throw new NoSuchElementException("Store not found");
+        } else {
+            store = resultStore.get();
+        }
+
+        List<OrderDetail> details = order.getProducts();
+        if (details == null || details.isEmpty()) {
+            throw new ConstraintViolationException("Empty order details", null);
+        }
+
+        if (order.getClientId() == null) {
+            throw new ConstraintViolationException("Order must have a clientId", null);
+        } else {
+            Optional<ClientEntity> resultClient = clientRepository.findById(order.getClientId());
+            if (!resultClient.isPresent()) {
+                throw new NoSuchElementException("Client not found");
+            } else {
+                client = resultClient.get();
+            }
+        }
+
+        OrderEntity newOrder = new OrderEntity(store, client);
+        newOrder = orderRepository.save(newOrder);
+        Order resultOrder = orderConverter.repositoryToApiModel(newOrder);
+
+        for(OrderDetail detail: details) {
+            long productId = detail != null && detail.getProductId() != null ? detail.getProductId() : 0;
+            Optional<ProductEntity> resultProduct = productRepository.findById(productId);
```

Comment:
So even if productId is null, you still query the database with productId=0? I think it's better to just skip the DB query if productId is null